### PR TITLE
fix(utils): add `TSDeclareFunction` to `functionTypeTypes`

### PR DESCRIPTION
**Pull request by abrahamguo**
## PR Checklist

- [ ] Addresses an existing open issue: fixes #000
- [ ] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

From https://github.com/typescript-eslint/typescript-eslint/pull/9479/commits/b47b88f9501ccbdac53a45e314ba69e403151b86, `TSDeclareFunction` was missing from `functionTypeTypes`.

### CODE REVIEW

File: packages/utils/src/ast-utils/predicates.ts
Comment (auvred) [inline packages/utils/src/ast-utils/predicates.ts:1]: [Testing] `isFunctionType` function is used in `no-redundant-type-constituents` rule. What do you think about adding a test case that works with `TSDeclareFunction` to the `no-redundant-type-constituents.test.ts`?

[playground](https://typescript-eslint.io/play/#ts=5.5.2&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVNQAoBKALnlRADcQZ4AfeAckRxyYCgOMBPABwSJE%2BALzwS8EQD4K1Wg2at2QA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1ukQBNYmvAIYtaRUrTIcmyfJXyxELVBkiJo0DtEjgwAXxB6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by kirkwaiblinger** [comment]
Would you mind filing an issue for this and/or fleshing out the description with the PR that this change is fixing up? The commit referenced says it's not on a branch within this repo, so it's hard to figure out what this is in reference to

**Comment by abrahamguo** [comment]
Yep, sorry! This came from #9479.

I agree with @auvred, I will add some unit tests to validate this change.

**Comment by abrahamguo** [comment]
In `no-redundant-type-constituents`, `never` is not allowed in unions.
However, an exception is for function return types, where it _is_ allowed in unions.
Because `TSDeclareFunction` was missing from `functionTypeTypes`, the rule wrongly _did_ report for `TSDeclareFunction`s but not other function types, as shown in @auvred's example:
```
declare function fn(): never | 'foo' // wrongly reports
type ffn = () => never | 'foo'
```

In this PR, it now correctly recognizes `TSDeclareFunction` as a function type, and now no longer reports.

File: packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts [hunk 1]
```diff
@@ -171,7 +170,7 @@ function describeLiteralTypeNode(typeNode: TSESTree.TypeNode): string {
 function isNodeInsideReturnType(node: TSESTree.TSUnionType): boolean {
   return !!(
     node.parent.type === AST_NODE_TYPES.TSTypeAnnotation &&
-    (isFunctionType(node.parent.parent) || isFunction(node.parent.parent))
+    isFunctionOrFunctionType(node.parent.parent)
```
Comment (abrahamguo) [inline packages/eslint-plugin/src/rules/no-redundant-type-constituents.ts:173]: `isFunctionOrFunctionType` simply combines the checks of the two functions into one

**Review by auvred — APPROVED** [review]

**Review by JoshuaKGoldberg — APPROVED** [review]
LGTM, thanks for the cleanup! 🙌

**Pull request merged**